Add new function `calculate_stats()`

mrcaseb opened the pull request:
This implements #445

mrcaseb commented:
It turned out that some stats are hard to compute using the playstats because some stats require information of other players, e.g. receiving_air_yards. 

The remaining stats probably need to be computed using pbp. I didn't start this yet

guga31bb commented on R/calculate_stats.R:
```diff
@@ -0,0 +1,263 @@
+  # seasons = 2023
+  # summary_level = "week"
+  # stat_type = "player"
+
```
Do we want a default summary level (season?) and stat type (player?)

And error checking for valid summary level / stat type provided?

mrcaseb replied:
Both argument defaults and inputs checks are done in lines 10+11

mrcaseb commented:
> What does this function do?

See here https://github.com/nflverse/nflfastR/blob/f59b2e672a3bb80b7ee570a63db0d070d0c46acb/NEWS.md?plain=1#L15

jacobakaye commented:
I installed the development version, and the old function 'calculate_player_stats' still worked while I got the error with the new function. 

calculate_stats(pbp)
Error in calculate_stats(pbp) : could not find function "calculate_stats"

mrcaseb commented:
@guga31bb and @tanho63 I have requested a review but honestly it's quite hard. Since it is a new function, it won't change anything at the moment. I think we can merge this and ask people to test it and look for possible bugs.

mrcaseb commented:
- Fumble recovery stats now count all recoveries (not only defense). That's necessary to count fumble recoveries of offensive players, esp. when they result in a TD, e.g. Trey McBride 2024, wk2. cc: @alecglen 
- added new variable `fumble_recovery_tds` to account for the above
- Changed variable names of fumble recovery and penalty stats to make clear they don't belong to defense only
- documented the changes and improved variable explainer in `nfl_stats_variables`

mrcaseb commented:
Changed variable names compared to the current stats approach are now documented here
https://github.com/nflverse/nflfastR/blob/cb83260032c33891c66fa2d33655337391ae6074/data-raw/nfl_stats_variables.R#L35-L65

This is the final call for review

TheMathNinja commented:
> Changed variable names compared to the current stats approach are now documented here
> 
> https://github.com/nflverse/nflfastR/blob/cb83260032c33891c66fa2d33655337391ae6074/data-raw/nfl_stats_variables.R#L35-L65
> 
> This is the final call for review

Does this mean the new version is erasing the difference between solo tackles and tackles with assist? Or will it maintain the two categories and also offer a combined variable?

mrcaseb commented:
> > Changed variable names compared to the current stats approach are now documented here
> 
> > 
> 
> > https://github.com/nflverse/nflfastR/blob/cb83260032c33891c66fa2d33655337391ae6074/data-raw/nfl_stats_variables.R#L35-L65
> 
> > 
> 
> > This is the final call for review
> 
> 
> 
> Does this mean the new version is erasing the difference between solo tackles and tackles with assist? Or will it maintain the two categories and also offer a combined variable?

It counts solo tackles, tackles with assist and tackle assists separately

TheMathNinja commented:
> > > Changed variable names compared to the current stats approach are now documented here
> > 
> > 
> > > 
> > 
> > 
> > > https://github.com/nflverse/nflfastR/blob/cb83260032c33891c66fa2d33655337391ae6074/data-raw/nfl_stats_variables.R#L35-L65
> > 
> > 
> > > 
> > 
> > 
> > > This is the final call for review
> > 
> > 
> > Does this mean the new version is erasing the difference between solo tackles and tackles with assist? Or will it maintain the two categories and also offer a combined variable?
> 
> It counts solo tackles, tackles with assist and tackle assists separately

Got it!

tanho63 commented:
```r
new_stats <- calculate_stats(2023, summary_level = "week", stat_type = "player")

old_stats <- dplyr::bind_rows(nflreadr::load_player_stats(2023,stat_type = "offense"),nflreadr::load_player_stats(2023,stat_type = "defense"),nflreadr::load_player_stats(2023,stat_type = "kicking"))

setdiff(names(old_stats), names(new_stats))
#> [1] "recent_team"                   "interceptions"                 "sacks"                        
#> [4] "sack_yards"                    "dakota"                        "def_tackles"                  
#> [7] "def_fumble_recovery_own"       "def_fumble_recovery_yards_own" "def_fumble_recovery_opp"      
#>[10] "def_fumble_recovery_yards_opp" "def_safety"                    "def_penalty"                  
#>[13] "def_penalty_yards"      

setdiff(names(new_stats),names(old_stats))
#> [1] "passing_interceptions"     "sacks_suffered"            "sack_yards_lost"           "passing_cpoe"             
#> [5] "def_safeties"              "misc_yards"                "fumble_recovery_own"       "fumble_recovery_yards_own"
#> [9] "fumble_recovery_opp"       "fumble_recovery_yards_opp" "fumble_recovery_tds"       "penalties"                
#> [13] "penalty_yards"             "punt_returns"              "punt_return_yards"         "kickoff_returns"          
#> [17] "kickoff_return_yards"     
```

I see most things have equivalents, should we document these somewhere e.g. what new fields are avail + what fields are no longer avail?

tanho63 commented on tests/testthat/test-calculate_stats.R:
```diff
@@ -0,0 +1,32 @@
+  expect_in(n3, var_names)
+  expect_in(n4, var_names)
+  expect_in(n5, var_names)
+
```
we could snapshot the schema and type as well e.g. make sure skimr::skim() is consistent for field names and values? 

skim <- skimr::skim(new_stats) %>% 
    as.data.frame() %>% 
    dplyr::transmute(variable = skim_variable, type = skim_type, n_missing = round(n_missing, -1))

expect_snapshot_value(skim)

mrcaseb replied:
done in e1f357fa0d55c96bb7944e5f1ed26edd0bee85d2

tanho63 reviewed (commented):
helluva job tackling this! performance not _superb_ but I fully support dplyr readability here over data.table performance/maintenance tradeoff since nflreadr will pull precalculated values 99% of the time. a few questions about field differences, I think.

mrcaseb commented:
> ```r
> new_stats <- calculate_stats(2023, summary_level = "week", stat_type = "player")
> 
> old_stats <- dplyr::bind_rows(nflreadr::load_player_stats(2023,stat_type = "offense"),nflreadr::load_player_stats(2023,stat_type = "defense"),nflreadr::load_player_stats(2023,stat_type = "kicking"))
> 
> setdiff(names(old_stats), names(new_stats))
> #> [1] "recent_team"                   "interceptions"                 "sacks"                        
> #> [4] "sack_yards"                    "dakota"                        "def_tackles"                  
> #> [7] "def_fumble_recovery_own"       "def_fumble_recovery_yards_own" "def_fumble_recovery_opp"      
> #>[10] "def_fumble_recovery_yards_opp" "def_safety"                    "def_penalty"                  
> #>[13] "def_penalty_yards"      
> 
> setdiff(names(new_stats),names(old_stats))
> #> [1] "passing_interceptions"     "sacks_suffered"            "sack_yards_lost"           "passing_cpoe"             
> #> [5] "def_safeties"              "misc_yards"                "fumble_recovery_own"       "fumble_recovery_yards_own"
> #> [9] "fumble_recovery_opp"       "fumble_recovery_yards_opp" "fumble_recovery_tds"       "penalties"                
> #> [13] "penalty_yards"             "punt_returns"              "punt_return_yards"         "kickoff_returns"          
> #> [17] "kickoff_return_yards"     
> ```
> 
> I see most things have equivalents, should we document these somewhere e.g. what new fields are avail + what fields are no longer avail? dakota and def_tackles for instance don't have direct equivs?

Done in the article that hosts the variables
